robots/bebop.h

```diff
@@ -0,0 +1,1188 @@
+    ~Bebop();
+    void addJoystick(HID::Joystick &joystick);
+    void connect();
+    void disconnect();
+
+    // speed limits
+    inline degree_t getMaximumTilt();
+    inline Limits<degree_t> &getTiltLimits();
+    inline meters_per_second_t getMaximumVerticalSpeed();
+    inline Limits<meters_per_second_t> &getVerticalSpeedLimits();
+    inline degrees_per_second_t getMaximumYawSpeed();
+    inline Limits<degrees_per_second_t> &getYawSpeedLimits();
+
+    // motor control
+    void takeOff();
+    void land();
+    void setPitch(const float pitch);
+    void setRoll(const float right);
+    void setVerticalSpeed(const float up);
+    void setYawSpeed(const float right);
+    void stopMoving();
+
+    // misc
+    VideoStream &getVideoStream();
+    void takePhoto();
+    void setFlightEventHandler(const FlightEventHandler);
+
+    // defaults
+    static constexpr auto DefaultMaximumTilt = 8_deg;
+    static constexpr auto DefaultMaximumYawSpeed = 100_deg_per_s;
+    static constexpr auto DefaultMaximumVerticalSpeed = 1_mps;
+
+private:
+    template<class UnitType>
+    class LimitValues
+    {
+    public:
+        UnitType m_UserMaximum;
+
+        inline void onChanged(const ARCONTROLLER_DICTIONARY_ELEMENT_t *dict,
+                              const char *currentKey,
+                              const char *minKey,
+                              const char *maxKey)
+        {
+            ARCONTROLLER_DICTIONARY_ELEMENT_t *elem = nullptr;
+            HASH_FIND_STR(dict, ARCONTROLLER_DICTIONARY_SINGLE_KEY, elem);
+            if (elem) {
+                ARCONTROLLER_DICTIONARY_ARG_t *arg = nullptr;
+
+                // get current value
+                HASH_FIND_STR(elem->arguments, currentKey, arg);
+                if (arg) {
+                    m_Current = units::make_unit<UnitType>(arg->value.Float);
+                }
+
+                // get min value
+                HASH_FIND_STR(elem->arguments, minKey, arg);
+                if (arg) {
+                    std::get<0>(m_Limits) = units::make_unit<UnitType>(arg->value.Float);
+                }
+
+                // get max value
+                HASH_FIND_STR(elem->arguments, maxKey, arg);
+                if (arg) {
+                    std::get<1>(m_Limits) = units::make_unit<UnitType>(arg->value.Float);
+                }
+
+                /*
+                 * Notify waiting threads. The if statement is necessary so that
+                 * we don't report the current maximum before we've had a chance
+                 * to set it to the user-preferred value.
+                 */
+                if (m_Current == m_UserMaximum) {
+                    m_Semaphore.notify();
+                }
+
+                // std::cout << "Max: " << m_Current << " ("
+                //           << std::get<0>(m_Limits) << ", "
+                //           << std::get<1>(m_Limits) << ")" << std::endl;
+            }
+        }
+
+        inline UnitType getCurrent()
```

Comment:
If you make the semaphore _mutable_ you can make these const

Reply:
Good point!